### ISSUE TITLE
"Client is currently disconnecting (32102)" upon Rapid AWSIotMqttManager disconnect()-then-connect()

### ISSUE THREAD

**Issue by ryamelv** [aws-amplify/aws-sdk-android#3036]
**Describe the bug**

If an app calls AWSIotMqttManager connect() quickly after a successful and complete call to disconnect() then a "Client is currently disconnecting (32102)" exception is thrown by AWSIotMqttManager.  A non-deterministic delay is required to avoid the exception and proceed with a successful connect().

**To Reproduce**

AWSIotMqttManager.connect()
...do things (subscribe, publish, etc.)...
AWSIotMqttManager.disconnect()
Wait for disconnect() to return.
Wait for AWSIotMqttClientStatusCallback.onStatusChanged() w/ status = ConnectionLost
AWSIotMqttManager.connect()

AWSIotMqttClientStatusCallback.onStatusChanged(0 receives status = Connecting, then status = ConnectionLost w/ exception:

```
Client is currently disconnecting (32102)
    at org.eclipse.paho.client.mqttv3.MqttAsyncClient.connect(MqttAsyncClient.java:729)
    at com.amazonaws.mobileconnectors.iot.AWSIotMqttManager.mqttConnect(AWSIotMqttManager.java:1164)
    at com.amazonaws.mobileconnectors.iot.AWSIotMqttManager.access$1100(AWSIotMqttManager.java:64)
    at com.amazonaws.mobileconnectors.iot.AWSIotMqttManager$1.run(AWSIotMqttManager.java:1013)
    at java.lang.Thread.run(Thread.java:923)
```

I have reproduced this in a basic/generic Android app that anyone can run in an emulator.  I happen to be running it in a Pixel 4 (API 30) emulator.

**Which AWS service(s) are affected?**

AWS SDK for Android (com.amazonaws.mobileconnectors.iot package)

**Expected behavior**

I expect that if disconnect() returns and AWSIotMqttClientStatusCallback.onStatusChanged() is called w/ status = ConnectionLost, then the same thread that called disconnect() should be able to call connect() immediately and not receive a "Client is currently disconnecting (32102)" exception (in its status/handler callback).

Rather, the connect() should succeed (assuming no other unrelated problems).  The app should not have to retry on this case in a loop, sleep() for a non-deterministic duration of time, etc.

**Screenshots**

There is no UI for this case.

**Environment Information (please complete the following information):**

 - AWS Android SDK Version: 2.50.1
 - Device: Pixel 4 (API 30) Emulator
 - Android Version: Android 11 (API 30)
 - Specific to simulators: No

**Additional context**

This seems similar to Issue #292 IoT MQTT Client Stuck Disconnecting.  But the resolution for that issue was to tell the OP to watch AWSIotMqttClientStatusCallback.onStatusChanged() for confirmation of the disconnect prior to calling connect(), and in this case I am doing that already.

Code inspection offers this theory:

[AWSIotMqttManager](https://aws-amplify.github.io/aws-sdk-android/docs/reference/index.html?com/amazonaws/mobileconnectors/iot/AWSIotMqttManager.html) uses [org.eclipse.paho.client.mqttv3.MqttAsyncClient](https://www.eclipse.org/paho/files/javadoc/org/eclipse/paho/client/mqttv3/MqttAsyncClient.html).

MqttAsyncClient uses [org.eclipse.paho.client.mqttv3.ClientComms](https://www.eclipse.org/paho/files/javadoc/org/eclipse/paho/client/mqttv3/internal/ClientComms.html).

AWSIotMqttManager maintains its own connection state member variable (“connectionState”).  Upon disconnect(), it passes the command on to MqttAsynClient which passes it on to ClientComms, the latter of which launches a background thread to do the full connection shutdown but then asynchronously reports “disconnected” back up the chain.

The result is that the application code’s invocation of disconnect() returns and even its status callback handler gets “Connection Lost” ahead of the ClientComms thread finishing its shutdown.

A subsequent call into AWSIotMqttManager.connect() uses its own (not fully correct) state member variable as a gate, which it gets past.  But once the connect command gets down to ClientComms it rejects it in its background thread because it knows its connection is still shutting down.

So then the application code’s status callback handler gets a MqttException.REASON_CODE_CLIENT_DISCONNECTING exception.

Perhaps AWSIotMqttManager needs to make sure the app using it doesn't get a "disconnect all done" callback ahead of ClientComms finishing all of its work.  Or, it needs to jettison the old ClientComms off to clean-up its stuff asynchronously and not block a new connect().  Or, the asynchronous nature of it all needs to be documented more clearly, and some facility provided by AWSIotMqttManager can make it more practical for the app to know when it's OK to call connect() again.

Or, I'm happy to be enlightened about what my app is doing wrong.  :)

The generic test app demonstrating this is a private repo.  I could grant access to limited folks that would benefit from seeing it to reproduce/address the issue.

**Comment by ryamelv** [comment]
Adding generic test app log, for a connect/subscribe/publish/disconnect/connect.  The first cycle (through the first disconnect() is as expected.  The start of the second cycle (three blank lines in the log output separate the two cycles) only gets to the connect() then gets the exception.

D/IotUserSession: iotCycle()...
I/IotUserSession: AWS IoT Device ID: <snip>
D/IotUserSession: CognitoUser...
D/IotUserSession: CognitoUser.getSession()...
...
I/CognitoAuthHandler: onSuccess(): <snip>
D/IotUserSession: CognitoCachingCredentialsProvider.setLogins()...
...
D/IotUserSession: AWSIotMqttManager.connect()...
D/IotUserSession: AWSIotMqttManager.connect().
D/MqttConnectHandler: waitForConnected()...
D/AWSIotMqttManager: MQTT broker: <snip>-ats.iot.us-east-1.amazonaws.com:443
...
D/CognitoCachingCredentia: Making a network call to fetch credentials.
V/com.amazonaws.request: Parsing service response JSON
D/com.amazonaws.request: CRC32Checksum = null
    content encoding = null
V/com.amazonaws.request: Done parsing service response
...
V/com.amazonaws.request: Parsing service response JSON
D/com.amazonaws.request: CRC32Checksum = null
    content encoding = null
V/com.amazonaws.request: Done parsing service response
...
D/AWSIotMqttManager: ready to do mqtt connect
I/AWSIotMqttManager: metrics collection is enabled, username: ?SDK=Android&Version=2.22.6
    resetting reconnect attempt and retry time
D/AWSIotMqttManager: Setting up Callback for MqttClient
D/MqttConnectHandler: onStatusChanged(): status = Connecting
I/AWSIotMqttManager: onSuccess: mqtt connection is successful.
D/MqttConnectHandler: onStatusChanged(): status = Connected
D/MqttConnectHandler: waitForConnected(): Connected
D/IotUserSession: AWSIotMqttManager.subscribeToTopic()...
D/IotUserSession: AWSIotMqttManager.subscribeToTopic().
D/MqttSubscribeHandler: waitForSubscription()...
W/MqttSubscribeHandler: onSuccess()
D/MqttSubscribeHandler: waitForSubscription(): Complete
D/IotUserSession: AWSIotMqttManager.publishString()...
D/IotUserSession: AWSIotMqttManager.publishString().
D/MqttPublishHandler: waitForPublish()...
I/AWSIotMqttManager: delivery is complete
D/MqttPublishHandler: statusChanged(): status = Success
D/MqttPublishHandler: waitForPublish(): Complete
D/MqttMessageHandler: waitForMessage()...
I/AWSIotMqttManager: message arrived on topic: <snip>/TopicAlpha
I/MqttMessageHandler: MQTT Recv: <snip>/TopicAlpha / Message #1
D/MqttMessageHandler: waitForMessage(): Received
D/IotUserSession: AWSIotMqttManager.unsubscribeTopic()...
    AWSIotMqttManager.unsubscribeTopic().
    AWSIotMqttManager.disconnect()...
D/MqttConnectHandler: onStatusChanged(): status = ConnectionLost
D/IotUserSession: AWSIotMqttManager.disconnect(): true
D/MqttConnectHandler: waitForDisconnected()...
    waitForDisconnected(): Disconnected
D/IotUserSession: iotCycle().



    iotCycle()...
I/IotUserSession: AWS IoT Device ID: <snip>
D/IotUserSession: AWSIotMqttManager.connect()...
    AWSIotMqttManager.connect().
D/MqttConnectHandler: waitForConnected()...
D/AWSIotMqttManager: MQTT broker: <snip>-ats.iot.us-east-1.amazonaws.com:443
D/AWSIotMqttManager: ready to do mqtt connect
I/AWSIotMqttManager: metrics collection is enabled, username: ?SDK=Android&Version=2.22.6
    resetting reconnect attempt and retry time
D/AWSIotMqttManager: Setting up Callback for MqttClient
D/MqttConnectHandler: onStatusChanged(): status = Connecting
D/MqttConnectHandler: onStatusChanged(): status = ConnectionLost
W/MqttConnectHandler: onStatusChanged()
    Client is currently disconnecting (32102)
        at org.eclipse.paho.client.mqttv3.MqttAsyncClient.connect(MqttAsyncClient.java:729)
        at com.amazonaws.mobileconnectors.iot.AWSIotMqttManager.mqttConnect(AWSIotMqttManager.java:1164)
        at com.amazonaws.mobileconnectors.iot.AWSIotMqttManager.access$1100(AWSIotMqttManager.java:64)
        at com.amazonaws.mobileconnectors.iot.AWSIotMqttManager$1.run(AWSIotMqttManager.java:1013)
        at java.lang.Thread.run(Thread.java:923)

**Comment by ryamelv** [comment]
Additional information:

In my generic test app, it reliably fails every-other connect() attempt in the loop as described previously.  But if I re-create the AWSIotMqttManager object for each loop iteration, then it succeeds every time as expected.

This is not a revolutionary insight.  :)  But it corroborates the afore described theory on the failure mode.

**Comment by eeatonaws** [comment]
Thank you for reporting this issue and providing such detailed information about the issue. A fix has been released as part of AWS SDK for Android 2.56.0.

**Comment by ryamelv** [comment]
Success!  I reverted my workaround (re-creating AWSIotMqttManager object for each loop iteration), re-produced the original problem, upgraded my AWS SDK for Android from v2.50.1 to v2.56.0, re-ran and observed normal operation (repeated loop iterations of connect/disconnect without the previous exception).  Without the workaround the app logs now look just like they did with the workaround.  Thank you for the fix!
